File: 프로그래머스/2/42587. 프로세스/프로세스.py
import heapq
from collections import deque

def solution(priorities, location):
    priority = []
    queue = deque([])
    
    for i, p in enumerate(priorities):
        heapq.heappush(priority, -p)
        queue.append((i, p))
    
    count = 0
    while priority:
        if priority[0] == -queue[0][1]:
            heapq.heappop(priority)
            curr_location, _ = queue.popleft()
            count += 1
            if curr_location == location:
                return count
        else:
            queue.append(queue.popleft())
            
    return 0

File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*;

class Task{
    int location;
    int priority;
    
    public Task(int location, int priority){
        this.location = location;
        this.priority = priority;
    }
}
class Solution {
    public int solution(int[] priorities, int location) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Comparator.reverseOrder());
        ArrayDeque<Task> readyQueue = new ArrayDeque<>();
        
        for(int i = 0; i < priorities.length; i++){
            pq.add(priorities[i]);
            readyQueue.addLast(new Task(i, priorities[i]));
        }
        
        int count = 0;
        while(!pq.isEmpty()){
            Task task = readyQueue.removeFirst();
            if(pq.peek() == task.priority){
                pq.poll();
                count++;
                if(task.location == location) return count;
            }else{
                readyQueue.addLast(task);
            }
        }
        
        return count;
    }
}